src/Agent.Listener/Configuration/FeatureFlagProvider.cs

```
@@ -0,0 +1,117 @@
+        /// <param name="serviceEndpoint">The service endpoint to get the feature flag status from.</param>
+        /// <param name="featureFlagName">The name of the feature flag to get the status of.</param>
+        /// <returns>The status of the feature flag.</returns>
+        public Task<FeatureFlagStatus> GetFeatureFlagAsync(IHostContext hostContext, ServiceEndpoint serviceEndpoint, string FeatureFlagName, ITraceWriter traceWriter);
+    }
+
+    public class FeatureFlagProvider : AgentService, IFeatureFlagProvider, IDisposable
+    {
+        private HttpClient _httpClient;
+
+        public FeatureFlagProvider()
+        {
+            
+        }
+
+        public FeatureFlagProvider(HttpClient client) 
+        {
+            _httpClient = client;
+        }
+
+        public async Task<FeatureFlagStatus> GetFeatureFlagAsync(IHostContext hostContext, ServiceEndpoint serviceEndpoint, string featureFlagName, ITraceWriter traceWriter)
+        {
+            ArgUtil.NotNull(serviceEndpoint, nameof(serviceEndpoint));
+            ArgUtil.NotNull(featureFlagName, nameof(featureFlagName));
+            using HttpClientHandler httpHandler = hostContext.CreateHttpClientHandler(); ;
+            if (_httpClient == null)
+            {
+                _httpClient = new HttpClient(httpHandler);
+            }
+            _httpClient.Timeout = TimeSpan.FromSeconds(10);
+            var accessTokenFF = serviceEndpoint?.Authorization?.Parameters["AccessToken"] ?? null;
+            var url = $"{serviceEndpoint.Url}_apis/FeatureFlags/{featureFlagName}";
+            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessTokenFF}");
+            var response = await _httpClient.GetAsync(url);
+            if (response.IsSuccessStatusCode)
+            {
+                try
+                {
+                    return await response.Content.ReadFromJsonAsync<FeatureFlagStatus>();
+                } catch(JsonException ex)
+                {
```

when do we expect to have a JsonException from the caller?  better not to handle the JsonException if we don't expect one.

---

@merlynomsft for example cases where this exception might be thrown:
1. FF not found in ADO side
2. authorization/token issue
3. network issues

---

2 or 3 I'd argue are not recoverable and job should fail to prevent unexpected behavior

for 1, we should implement Try pattern referenced below.  
https://github.com/microsoft/azure-pipelines-agent/pull/4415#discussion_r1322002240

but ideally we'd handle the error very narrowly .  *Assuming we can only handle missing FFs by catching*, we shouldn't assume all JsonExceptions are missing feature flag.  better to check if there a return value, messge or field on JsonException we can check for missing FF.  For everything else, re-throw the exception using "throw;"